src/Build/Graph/ProjectGraph.cs

```diff
@@ -75,6 +76,16 @@ public sealed class ProjectGraph
 
         public IReadOnlyCollection<ProjectGraphNode> GraphRoots { get; }
 
+        /// <summary>
+        ///     Gets a distinct enumeration of all of the environment variables reads during evaluation.
+        /// </summary>
+        public IEnumerable<string> EnvironmentVariableReads => _projectNodesTopologicallySorted?.Value.Select(p => p.ProjectInstance).SelectMany(pi => pi.EnvironmentVariableReads).Distinct();
+
+        /// <summary>
+        ///     Get a distinct enumeration of all the uninitialized property reads during evaluation.
+        /// </summary>
+        public IEnumerable<string> UninitializedPropertyReads => _projectNodesTopologicallySorted?.Value.Select(p => p.ProjectInstance).SelectMany(pi => pi.UninitializedPropertyReads).Distinct();
```

Comment:
>public [](start = 8, length = 6)

I'd add another property 'EnvironmentVariablesImpactingBuild' (or similar), that is the union of these two. At least from a buildxl point of view, that's really what we want to know, and present vs. absent variables are not really important

Reply:
Added.